controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,430 @@
+
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	// Create an empty hostFirmwareSetting with same name and namespace as host
+	hfs.ObjectMeta = metav1.ObjectMeta{
+		Name:      info.namespaceName.Name,
+		Namespace: info.namespaceName.Namespace}
+	hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	hfs.Spec.Settings = make(metal3v1alpha1.DesiredSettingsMap)
+	hfs.Annotations = make(map[string]string)
+
+	// Get the baremetalhost in this namespace in order to retrieve the provisioner ID
+	// If host is not available still create the hfs and can get the ID on subsequent updates
+	info.hfs = hfs
+	if err = r.getProvisionInfoFromBMH(info); err != nil {
+		info.log.Info("could not get provisioner ID", "error", err)
+	}
+
+	if err = r.Create(context.TODO(), hfs); err != nil {
+		return errors.Wrap(err, "could not create hostFirmwareSettings resource")
+	}
+	info.log.Info("created new hostFirmwareSettings resource")
+
+	return nil
+}
+
+// For the baremetalhost that matches the name/namespace get the provisioner ID and hardware info
+func (r *HostFirmwareSettingsReconciler) getProvisionInfoFromBMH(info *rInfo) (err error) {
+
+	bmh := &metal3v1alpha1.BareMetalHost{}
+	if err = r.Get(context.TODO(), info.namespaceName, bmh); err == nil {
+
+		// Store info in annotation that will persist if settings is cleared during pivot
+		info.hfs.Annotations[metal3v1alpha1.ProvisionerIdAnnotation] = bmh.Status.Provisioning.ID
+		// Hardware info is optional
+		if bmh.Status.HardwareDetails != nil {
+			info.hfs.Annotations[metal3v1alpha1.HardwareVendorAnnotation] = bmh.Status.HardwareDetails.SystemVendor.Manufacturer
+			info.hfs.Annotations[metal3v1alpha1.HardwareModelAnnotation] = bmh.Status.HardwareDetails.SystemVendor.ProductName
+		}
+	} else {
+
+		return errors.Wrap(err, "could not get baremetalhost for provisioner id")
+	}
+
+	return nil
+}
+
+// EventHandler for updates to both the baremetalhost and hostfirmwarettings
+func (r *HostFirmwareSettingsReconciler) updateEventHandler(e event.UpdateEvent) bool {
+
+	r.Log.Info("hostfirmwaresettings in event handler")
+
+	// If this is a baremetalhost only return true on certain state transitions
+	var oldState metal3v1alpha1.ProvisioningState = metal3v1alpha1.StateNone
+	var newState metal3v1alpha1.ProvisioningState = metal3v1alpha1.StateNone
+	oldHost, oldHostExists := e.ObjectOld.(*metal3v1alpha1.BareMetalHost)
+	if oldHostExists {
+		oldState = oldHost.Status.Provisioning.State
+	}
+	newHost, newHostExists := e.ObjectNew.(*metal3v1alpha1.BareMetalHost)
+	if newHostExists {
+		newState = newHost.Status.Provisioning.State
+	}
+	if oldHostExists || newHostExists {
+		// The data only needs to be retrieved from Ironic after the node has been moved to
+		// ready and when the node is fully provisionied, its not necessary on every state
+		// change as the data from Ironic will be the same.
+		if newState == metal3v1alpha1.StateReady && oldState == metal3v1alpha1.StatePreparing {
```

Comment:
That's not clear to me. What's the reason to reconcile when the BMH goes into `Ready` from `Preparing`?

Reply:
I updated the comment to explain this a bit better. We don't need to retrieve data from ironic on every state transition but we do need to get it after cleaning as the settings could have changed (i.e. from Preparing to Ready).